<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ cancerCellNet</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Function reference" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cancerCellNet</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="addRandToSampTab.html">addRandToSampTab()</a></code> </p>
        </td>
        <td><p>Add Random Profiles to Sample Table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="avgGeneCat.html">avgGeneCat()</a></code> </p>
        </td>
        <td><p>Average Category Gene Expressions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="barPlotGene.html">barPlotGene()</a></code> </p>
        </td>
        <td><p>Plot Specific Gene Expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="broadClass_ensemblePredict.html">broadClass_ensemblePredict()</a></code> </p>
        </td>
        <td><p>Predict query using an ensemble of broad class classifier</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="broadClass_ensembleTrain.html">broadClass_ensembleTrain()</a></code> </p>
        </td>
        <td><p>Broad Class Ensemble Training (not commonly used)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="broadClass_predict.html">broadClass_predict()</a></code> </p>
        </td>
        <td><p>Predict query using broad class classifier</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="broadClass_train.html">broadClass_train()</a></code> </p>
        </td>
        <td><p>Broad Class Training</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ccn_classAssess.html">ccn_classAssess()</a></code> </p>
        </td>
        <td><p>Assessment of Classifier</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ccn_hmClass.html">ccn_hmClass()</a></code> </p>
        </td>
        <td><p>Heatmap of the classification result</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cn_clPerf.html">cn_clPerf()</a></code> </p>
        </td>
        <td><p>run cn_clPerf across thresholds</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cn_correctZmat.html">cn_correctZmat()</a></code> </p>
        </td>
        <td><p>make Inf and -Inf values sensible</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cn_eval.html">cn_eval()</a></code> </p>
        </td>
        <td><p>determine performance of classification at given threshold</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cn_HmClass.html">cn_HmClass()</a></code> </p>
        </td>
        <td><p>Simple heatmap of the classification result</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compareGenePairs.html">compareGenePairs()</a></code> </p>
        </td>
        <td><p>Compile genePairs comparison matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="conversionList.html">conversionList</a></code> </p>
        </td>
        <td><p>List of Conversion dataframe</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="downSampleW.html">downSampleW()</a></code> </p>
        </td>
        <td><p>Weighted subtraction from mapped reades</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findBestPairs.html">findBestPairs()</a></code> </p>
        </td>
        <td><p>Find best pairs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findClassyGenes.html">findClassyGenes()</a></code> </p>
        </td>
        <td><p>Find Classifier-Worthy Gene Candidates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="find_genes_byGo.html">find_genes_byGo()</a></code> </p>
        </td>
        <td><p>find transcript factors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_expr_sum.html">gene_expr_sum()</a></code> </p>
        </td>
        <td><p>Sum transcription expression estimates to gene-level expression measures</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEP_makeMean.html">GEP_makeMean()</a></code> </p>
        </td>
        <td><p>row average (or median) based on groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getClassGenes.html">getClassGenes()</a></code> </p>
        </td>
        <td><p>Find Classy Genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gnrAll.html">gnrAll()</a></code> </p>
        </td>
        <td><p>Find Genes in Clusters compare to other categories</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gnrBP.html">gnrBP()</a></code> </p>
        </td>
        <td><p>find best pairs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hm_gpa_sel.html">hm_gpa_sel()</a></code> </p>
        </td>
        <td><p>Heatmap of the genes and groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeClassifier.html">makeClassifier()</a></code> </p>
        </td>
        <td><p>Make Classifier</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeGeneCompareTab.html">makeGeneCompareTab()</a></code> </p>
        </td>
        <td><p>Make Gene Comparison Table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makePairTab.html">makePairTab()</a></code> </p>
        </td>
        <td><p>Make the pair tabs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeSubClassifier.html">makeSubClassifier()</a></code> </p>
        </td>
        <td><p>Make Classifier tailored for subclassifier</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ModifiedRandomize.html">ModifiedRandomize()</a></code> </p>
        </td>
        <td><p>Randomize the data matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pair_transform.html">pair_transform()</a></code> </p>
        </td>
        <td><p>Gene to Gene comparison</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotGeneComparison.html">plotGeneComparison()</a></code> </p>
        </td>
        <td><p>Gene expression plotting</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_class_PRs.html">plot_class_PRs()</a></code> </p>
        </td>
        <td><p>Plot precision recall curve per category</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ptGetTop.html">ptGetTop()</a></code> </p>
        </td>
        <td><p>Find the best gene pairs for training</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ptSmall.html">ptSmall()</a></code> </p>
        </td>
        <td><p>Pair Transform on small scale</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="query_transform.html">query_transform()</a></code> </p>
        </td>
        <td><p>Make complete gene-to-gene comparison</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="randomize.html">randomize()</a></code> </p>
        </td>
        <td><p>Randomize the data matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rf_classPredict.html">rf_classPredict()</a></code> </p>
        </td>
        <td><p>Classify Query Samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc_compAlpha.html">sc_compAlpha()</a></code> </p>
        </td>
        <td><p>Compute Alpha</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc_compMu.html">sc_compMu()</a></code> </p>
        </td>
        <td><p>compute Mu</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc_cov.html">sc_cov()</a></code> </p>
        </td>
        <td><p>Compute Coefficient of Variation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc_fano.html">sc_fano()</a></code> </p>
        </td>
        <td><p>Compute Fano Factor</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc_filterCells.html">sc_filterCells()</a></code> </p>
        </td>
        <td><p>find cells that pass criteria</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc_filterGenes.html">sc_filterGenes()</a></code> </p>
        </td>
        <td><p>Find Genes that Pass Criteria</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc_sampR_to_pattern.html">sc_sampR_to_pattern()</a></code> </p>
        </td>
        <td><p>Create Template</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc_statTab.html">sc_statTab()</a></code> </p>
        </td>
        <td><p>Generate a Stats Table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc_testPattern.html">sc_testPattern()</a></code> </p>
        </td>
        <td><p>Template Matching</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sdGeneCat.html">sdGeneCat()</a></code> </p>
        </td>
        <td><p>Standard Deviation Category Gene Expressions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="splitCommon.html">splitCommon()</a></code> </p>
        </td>
        <td><p>Split Sample Table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subClassQuery_transform.html">subClassQuery_transform()</a></code> </p>
        </td>
        <td><p>Sub-class Query Transform</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subClass_hmClass.html">subClass_hmClass()</a></code> </p>
        </td>
        <td><p>Detailed Complex heatmap for subclass classification</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subClass_predict.html">subClass_predict()</a></code> </p>
        </td>
        <td><p>Predict sub class query using broad class and subclass classifier (newer version)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subClass_train.html">subClass_train()</a></code> </p>
        </td>
        <td><p>Sub-Class Training (new and improved version) with merged rand category</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="trans_prop.html">trans_prop()</a></code> </p>
        </td>
        <td><p>Log scaled downsampled expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="utils_convertLongTab.html">utils_convertLongTab()</a></code> </p>
        </td>
        <td><p>Convert Long Expression Table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="utils_convertToGeneSymbols.html">utils_convertToGeneSymbols()</a></code> </p>
        </td>
        <td><p>Gene Symbol Conversion</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="utils_loadObject.html">utils_loadObject()</a></code> </p>
        </td>
        <td><p>Load an R object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="utils_matchSampTab.html">utils_matchSampTab()</a></code> </p>
        </td>
        <td><p>Match Sample Table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="utils_myDate.html">utils_myDate()</a></code> </p>
        </td>
        <td><p>print date</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="utils_myDist.html">utils_myDist()</a></code> </p>
        </td>
        <td><p>1-PCC distance</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="utils_stripwhite.html">utils_stripwhite()</a></code> </p>
        </td>
        <td><p>strip whitespace from a string</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="utils_strip_fname.html">utils_strip_fname()</a></code> </p>
        </td>
        <td><p>reduces full path to filename</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="weighted_down.html">weighted_down()</a></code> </p>
        </td>
        <td><p>Down Sample Gene Expressions</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Patrick Cahan.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

